tests/test_directives.py

```diff
@@ -0,0 +1,270 @@
+        assert NP.finalize(2, dict_directives) == 2
+        dict_directives['nranks'] = 2
+        dict_directives['omp_num_threads'] = 4
+        assert NP.finalize(2, dict_directives) == 8
+        dict_directives['nranks'] = lambda x: x**2
+        assert NP.finalize(2, dict_directives) == 2
+
+    def test_manual_item_default(self, product_directive):
+        assert product_directive.default == 10
+
+    def test_manual_item_validation(self, product_directive):
+        try:
+            val = product_directive.validation(product_directive.default)
+            assert product_directive.default == val
+        except Exception:
+            raise ValueError("Invalid default value for product_directive")
+        assert product_directive.default == product_directive.validation(product_directive.default)
+        assert product_directive.validation(20) == 20
+        with pytest.raises(ValueError):
+            product_directive.validation(0)
+
+    def test_manual_item_serial(self, product_directive):
+        product_directive.serial(10, 20) == 10
+
+    def test_manual_item_parallel(self, product_directive):
+        product_directive.parallel(10, 20) == 10
+
+    def test_manual_item_finalize(self, product_directive):
+        asset_dict = {'free': False, 'discount': 5}
+        assert product_directive.finalize(50, asset_dict) == 45
+        asset_dict['free'] = True
+        assert product_directive.finalize(50, asset_dict) == 0
+
+
+class TestDirectives:
+    """
+    Tests for Directives Class
+    """
+    def test_get_directive(self, available_directives_list):
+        directives = Directives(available_directives_list)
+        for item in available_directives_list:
+            assert directives[item.name] == item.default
+
+    def test_add_directive(self, available_directives_list):
+        directives = Directives(available_directives_list[:-1])
+        directives._add_directive(PROCESS_FRACTION)
+        assert directives[PROCESS_FRACTION.name] == PROCESS_FRACTION.default
+        with pytest.raises(TypeError):
+            directives._add_directive('Test')
+        with pytest.raises(ValueError):
+            directives._add_directive(PROCESS_FRACTION)
+
+    def test_set_defined_directive(self, available_directives_list):
+        directives = Directives(available_directives_list)
+        directives._set_defined_directive(NP.name, 10)
+        assert directives[NP.name] == 10
+
+    def test_set_defined_directive_invalid(self, available_directives_list):
+        directives = Directives(available_directives_list)
+        with pytest.raises(ValueError):
+            directives._set_defined_directive(NP.name, 0)
+
+    def test_set_undefined_directive(self, available_directives_list, product_directive):
+        directives = Directives(available_directives_list)
+        with pytest.raises(Exception):
+            directives._set_defined_directive(product_directive.name, 0)
```

Comment:
No need to invoke the `product_directive` here. Just testing against some random string is fine.

Reply:
Noted.